src/shared/tpm2-util.c

```diff
@@ -307,55 +317,206 @@ static int tpm2_credit_random(ESYS_CONTEXT *c) {
         return 0;
 }
 
-static int tpm2_make_primary(
-                ESYS_CONTEXT *c,
-                ESYS_TR *ret_primary,
-                TPMI_ALG_PUBLIC alg,
-                TPMI_ALG_PUBLIC *ret_alg) {
+const TPM2B_PUBLIC *tpm2_get_primary_template(Tpm2SRKTemplateFlags flags) {
 
-        static const TPM2B_SENSITIVE_CREATE primary_sensitive = {};
-        static const TPM2B_PUBLIC primary_template_ecc = {
-                .size = sizeof(TPMT_PUBLIC),
-                .publicArea = {
-                        .type = TPM2_ALG_ECC,
-                        .nameAlg = TPM2_ALG_SHA256,
-                        .objectAttributes = TPMA_OBJECT_RESTRICTED|TPMA_OBJECT_DECRYPT|TPMA_OBJECT_FIXEDTPM|TPMA_OBJECT_FIXEDPARENT|TPMA_OBJECT_SENSITIVEDATAORIGIN|TPMA_OBJECT_USERWITHAUTH,
-                        .parameters.eccDetail = {
-                                .symmetric = {
-                                        .algorithm = TPM2_ALG_AES,
-                                        .keyBits.aes = 128,
-                                        .mode.aes = TPM2_ALG_CFB,
+        /*
+         * Set up array so flags can be used directly as an input.
+         *
+         * Templates for SRK come from the spec:
+         *   - https://trustedcomputinggroup.org/wp-content/uploads/TCG-TPM-v2.0-Provisioning-Guidance-Published-v1r1.pdf
+         *
+         * However, note their is some lore here. On Linux, the SRK has it's unique field set to size 0 and
+         * on Windows the SRK has their unique data set to keyLen in bytes of zeros.
+         */
+        assert(flags >= 0);
+        assert(flags <= _TPM2_SRC_TEMPLATE_MAX);
+
+        static const TPM2B_PUBLIC templ[] = {
+                /* index 0 RSA old */
+                {
```

Comment:
i'd make this more explicit, i.e. add here:

```c
static const TPM2B_PUBLIC templ[_TPM2_SRC_TEMPLATE_MAX] = {
        [0] = { 
                …
        },
        [TPM2_SRK_TEMPLATE_ECC] = {
                …
        },
        [TPM2_SRK_TEMPLATE_NEW_STYLE] = { 
                …
        },
        [TPM2_SRK_TEMPLATE_NEW_STYLE|TPM2_SRK_TEMPLATE_ECC] = {
                …
        }
};

Reply:
This is 3, when the len is 4, so either I need to add one to `_TPM2_SRC_TEMPLATE_MAX` or 1 to the array definition, do you have a preference?

Reply:
For now I just did a `+ 1` at array definitions and performed less than or equal tests on incoming flags to the constant.

Reply:
yeah, that works.